src/v/cluster/cloud_metadata/manifest_downloads.cc

```diff
@@ -145,4 +148,88 @@ ss::future<std::list<ss::sstring>> list_orphaned_by_manifest(
     co_return ret;
 }
 
+ss::future<cluster_manifest_result> download_highest_manifest_in_bucket(
+  cloud_storage::remote& remote,
+  const cloud_storage_clients::bucket_name& bucket,
+  retry_chain_node& retry_node) {
+    // Look for unique cluster UUIDs for which we have metadata.
+    constexpr auto cluster_prefix = "cluster_metadata/";
+    vlog(clusterlog.trace, "Listing objects with prefix {}", cluster_prefix);
+    auto list_res = co_await remote.list_objects(
+      bucket,
+      retry_node,
+      cloud_storage_clients::object_key(cluster_prefix),
+      std::nullopt);
+    if (list_res.has_error()) {
+        vlog(clusterlog.debug, "Error downloading manifest", list_res.error());
+        co_return error_outcome::list_failed;
+    }
+    // Examine all cluster metadata in this bucket.
+    auto& cluster_metadata_items = list_res.value().contents;
+    if (cluster_metadata_items.empty()) {
+        vlog(clusterlog.debug, "No manifests found in bucket {}", bucket());
+        co_return error_outcome::no_matching_metadata;
+    }
+
+    // Look through those that look like cluster metadata manifests and find
+    // the one with the highest metadata ID. This will be the returned to the
+    // caller.
+    model::cluster_uuid uuid_with_highest_meta_id{};
+    cluster_metadata_id highest_meta_id{};
+    for (const auto& item : cluster_metadata_items) {
+        std::smatch matches;
+        std::string k = item.key;
+        const auto matches_manifest_expr = std::regex_match(
+          k.cbegin(), k.cend(), matches, cluster_metadata_manifest_expr);
+        if (!matches_manifest_expr) {
+            continue;
+        }
+        const auto& cluster_uuid_str = matches[1].str();
```

Comment:
nit: I'd assert on the size of `matches`

Reply:
Done